### DISCUSSION TITLE
new sync with filter parameters deleting previous records synched.

### DISCUSSION THREAD

**Discussion by sojan1** [Mimetis/Dotmim.Sync#1108] [Q&A]
if there is a delete cascade applied on a field, DMS is deleting the previous record synced with filter from the parent table.
eg:  in the example below, it is deleting from the 'project' table, when there is new sync with new filter.

https://github.com/sojan1/SyncInstaller

first sync
var parameters2 = new SyncParameters
{
{ "CLNT_ID","D1" },
{ "PROJ_ID","MASTER" }
};

second sync
var parameters2 = new SyncParameters
{
{ "CLNT_ID","D1" },
{ "PROJ_ID","Demo 2a" }
};

and the record gets removed from the hub, if  synced 'MASTER' again.

**Comment by Mimetis** [comment]
disable cascade, I guess

**Comment by sojan1** [comment]
Hi
Is that an issue or is that how it is planned ? 
Will you also not recommend update cascade?

We are using cascading on the mobile app to delete records. 
This will allow us to delete records from all sub tables without using separate sql statements to delete from each table.

**Comment by sojan1** [comment]
On every synchronisation, it is deleting and then re-downloading all the data in all tables. 

I have confirmed this by giving a wrong 'PROJ_ID' parameter, which removed existing rows from 'PROJECT' table and '_desx' Table, but there was data in the _clnt table (_clnt table(parent) is without filters, and don't have PROJ_ID).

deleting the rows in _clnt table will definitely remove data from other tables as there is cascades on all there tables and thus we dont see any data after applying new filter for the sync.

![image](https://github.com/Mimetis/Dotmim.Sync/assets/111710684/2493d6d1-0b12-4d73-a29a-9292625a517d)

**Comment by sojan1** [comment]
**The issue i have mentioned above is demonstrated on the the video link added.** Issue exists from DMS 0.9.8 onwards.
 https://sojan-geroc.tinytake.com/msc/ODc1MTg1M18yMjE2MDk5Mw

We intend to provision the server and client databases to synchronise data for specific projects only.
Each client database may contain one or mode projects at a given time.
Running a sync for a specific project should not affect any data in other projects.
Our goal is to provision a single scope that can be used across multiple projects simply by changing the filter values to be used.
When we ask to sync ‘Project 1’ we only want to make changes to ‘Project 1’.
This worked as expected with 0.9.7.
Since 0.9.8 data in other projects is being ‘deleted’.

**Reply by Mimetis** [reply]
I tried with the last beta version [1.0.0-beta-1517](https://www.nuget.org/packages/Dotmim.Sync.Core/1.0.0-beta-1517) and it's working fine on my setup:

![image](https://github.com/Mimetis/Dotmim.Sync/assets/4592555/02109aa6-c63b-4e67-b37b-ab6c5342c8b0)

**Comment by sojan1** [comment]
Hi , Can you kindly check if the cascade is applied on the sqlite database Or the DisableConstraintsOnApplyChanges is not enabled on sync.  Because i can see the same issue on 1.0.0-beta-1517

i have not deleted any record from _clnt table, but it shows 4 inserts and 4 deletes.
![image](https://github.com/Mimetis/Dotmim.Sync/assets/111710684/da118ee6-7fcc-41b4-bec6-2b4f9e6abde1)

**Comment by sojan1** [comment]
@Mimetis while looking through the code i found this link  https://github.com/Mimetis/Dotmim.Sync/discussions/453#discussioncomment-380530

Was the DisableConstraintsOnApplyChanges -> 'true' by default in  0.9.7 version? 
Does the post say that DMS will delete existing rows every time and reinsert if DisableConstraintsOnApplyChanges is set to true. 

it still function in the same manner ? or @gentledepp 's suggession have already applied?
![image](https://github.com/Mimetis/Dotmim.Sync/assets/111710684/fdf8bb6c-c722-466a-b15b-f1ba5f988b0e)


thanks

**Reply by Mimetis** [reply]
DisableConstraintsOnApplyChanges  is here to disable constraints when applying changes, only

**Reply by sojan1** [reply]
Just to confirm,  **As said in the #453 will DMS delete the existing rows and re-insert every time based on the scope_id?** 


Will it not replace the old record with the new record if the scope_id is same for each new scope filter? 
(as we are deleting and inserting based on new scope, and time stamp)

**Is the sync_scope_hash used in the 'where' clause,**  as it is different for every new scopes? 
![image](https://github.com/Mimetis/Dotmim.Sync/assets/111710684/85edab18-8d16-4ab4-a59c-cb2f30e3d05c)

Do you foresee any issue if DisableConstraintsOnApplyChanges is set to 'false'?

**Reply by Mimetis** [reply]
I don't understand
hash, is just the hash of the column parameters, nothing else

**Reply by sojan1** [reply]
i have synchronised sqlite db with hub, using 3 different filters, one filter after other.

Now i have 3 scopes with same scope_id and scope name on sqlite scope_info_client table as shown below 
(with 3 different filter parameters)
![image](https://github.com/Mimetis/Dotmim.Sync/assets/111710684/1873de3c-ca9c-4d50-952a-422eb620f8be)

**Which one in the above list of rows would have applied by DMS in the 3rd attempt? 
The last row or all the rows?** 


if it is only the last row that is applied, based on the timestamp and scope_id,
And if DMS is deleting the records based on the scope_id, before re-inserting as said in #453  (with cascade applied on tables). 
Will it not remove the previous records from the sqlite child tables, after new sync?

**Reply by richard-sisson** [reply]
@Mimetis - This post seems to be going off-track from the original question.

Before I get started, thanks for your efforts!  I'm really glad to see this project as we're needing to migrate from an existing MSSyncFramework setup. 

We did all of our viability testing using the 0.9.7 version, testing all of the features that are critical to us, including the reuse of scopes with different filter parameters (in our case these are projects).  Essentially everything passed using 0.9.7, but we have noticed a significant change in behaviour from 0.9.8 onwards.

From 0.9.8 we see the following:
- provision and initial sync parameters (P0) - objects are added to the databases and data is synced from the server to client database
- change sync parameters (P1) - data is synced from the server to client database, P0 data is **deleted from the client**
- change sync parameters (P2) - data is synced from the server to client database, P1 data is **deleted from the client**, P0 is **deleted from the server**

In 0.9.7 data was not deleted.  Why is it now deleting now?  Is this by design?

For us, data should only be deleted if a user has explicitly deleted data in a specific project.

Thanks
Richard

**Reply by Mimetis** [reply]
Hello Richard.
As I don't understand exactly what you need, the best solution for me would be to have a demo that I can use to reproduce your error.
Please use a demo script simple, explained, with the error at the end, or at least the result you want and that you don't have.

In your case, P0 and P1 should not be deleted after P2

**Reply by Mimetis** [reply]
Here is an example I created based on the [AdventureWorks](https://github.com/Mimetis/Dotmim.Sync/blob/master/CreateAdventureWorks.sql) sample:

I'm using the latest **DMS 1.0.0 beta** version.

``` cs
var setup = new SyncSetup("Product");
var options = new SyncOptions();

var agent = new SyncAgent(clientProvider, serverProvider, options);

setup.Filters.Add("Product", "Color");

// P0 : Color: "Red"
var p0 = new SyncParameters { { "Color", "Red" } };
var resultP0 = await agent.SynchronizeAsync(setup, p0);
Console.WriteLine(resultP0);

var clientConnection = clientProvider.CreateConnection();
var productRows = await clientProvider.GetDatabaseBuilder().GetTableAsync("Product", null, clientConnection);
Console.WriteLine($"rows: {productRows.Rows.Count}");

// P1 : Color: "Black"
var p1 = new SyncParameters { { "Color", "Black" } };
var resultP1 = await agent.SynchronizeAsync(setup, p1);
Console.WriteLine(resultP1);

productRows = await clientProvider.GetDatabaseBuilder().GetTableAsync("Product", null, clientConnection);
Console.WriteLine($"rows: {productRows.Rows.Count}");

// P2 : Color: "Black"
var p2 = new SyncParameters { { "Color", "Silver" } };
var resultP2 = await agent.SynchronizeAsync(setup, p2);
Console.WriteLine(resultP2);

productRows = await clientProvider.GetDatabaseBuilder().GetTableAsync("Product", null, clientConnection);
Console.WriteLine($"rows: {productRows.Rows.Count}");

```

And as expected, the results: 

``` 
Synchronization done.
        Total changes  uploaded: 0
        Total changes  downloaded: 38
        Total changes  applied on client: 38
        Total changes  applied on server: 0
        Total changes  failed to apply on client: 0
        Total changes  failed to apply on server: 0
        Total resolved conflicts: 0
        Total duration :00.00:01.679
rows: 38
Synchronization done.
        Total changes  uploaded: 0
        Total changes  downloaded: 89
        Total changes  applied on client: 89
        Total changes  applied on server: 0
        Total changes  failed to apply on client: 0
        Total changes  failed to apply on server: 0
        Total resolved conflicts: 0
        Total duration :00.00:00.057
rows: 127
Synchronization done.
        Total changes  uploaded: 0
        Total changes  downloaded: 36
        Total changes  applied on client: 36
        Total changes  applied on server: 0
        Total changes  failed to apply on client: 0
        Total changes  failed to apply on server: 0
        Total resolved conflicts: 0
        Total duration :00.00:00.034
rows: 163
```

And of course, the 3 scopes created, based on the 3 different parameter value:

![image](https://github.com/Mimetis/Dotmim.Sync/assets/4592555/e8805645-87a9-4828-9111-ea60027d92d4)

**Reply by richard-sisson** [reply]
Can you please check the details as indicated in this example image to see if there any deletes attempted:
![image](https://github.com/Mimetis/Dotmim.Sync/assets/10456433/0ac15483-c779-49ad-bd73-30798c086628)

These were not showing in the summary data for us.

If you still do not see anything I will setup a separate test from the one that Sojan has been running.

Thanks
Richard

**Reply by Mimetis** [reply]
Hi, Richard, 

As I mentioned many times, please, make an example I can reproduce.
Something like what I did in the previous comment.

I'm sorry, but this is the last time I'm trying to help you.

IF you can't provide a simple example (as I did in the previous comment) I will not help you.
I'm sorry if it's a little bit rude, but I asked it many times, and you didn't listen.

If you can't make this effort, my advice to you is to search for another solution, preferably a paid solution with a full customer support, not an open source project

Open Source Must be a Joint Effort

**Reply by sojan1** [reply]
Hi @Mimetis,

Apoligize for any misunderstanding occured.

We are using **version 1.0.0. Beta**
We are using the same **Adventureworks** database provided in the above post.
Apply CASCADE on SQLite 'Product' table-> ProductCategoryID to see the issue.

Full Source Code - > https://github.com/sojan1/Dotmim.Sync.SampleConsole/tree/master
    

        var serverProvider = new SqlSyncProvider(DBHelper.GetDatabaseConnectionString(serverDbName));
        SqliteSyncProvider clientProvider = new SqliteSyncProvider("advworks.db");

        var setup = new SyncSetup("ProductCategory", "Product");
        var options = new SyncOptions();

        var agent = new SyncAgent(clientProvider, serverProvider, options);

        var proFilter = new SetupFilter("Product");
        proFilter.AddParameter("Color", "Product", false);
        proFilter.AddWhere("Color", "Product", "Color");
        proFilter.AddParameter("ProductCategoryID", "Product", false);
        proFilter.AddWhere("ProductCategoryID", "Product", "ProductCategoryID");
        setup.Filters.Add(proFilter);


        var p0 = new SyncParameters { { "Color", "Red" }, { "ProductCategoryID", "5515f857-075b-4f9a-87b7-43b4997077b3" } };
        var resultP0 = await agent.SynchronizeAsync(setup, p0);
        Console.WriteLine(resultP0);
        var clientConnection = clientProvider.CreateConnection();
        var productRows = await clientProvider.GetDatabaseBuilder().GetTableAsync("ProductCategory", null, clientConnection);
        Console.WriteLine($"rows: {productRows.Rows.Count}");

 
        var p1 = new SyncParameters { { "Color", "Black" }, { "ProductCategoryID", "5515f857-075b-4f9a-87b7-43b4997077b3" } };
        var resultP1 = await agent.SynchronizeAsync(setup, p1);
        Console.WriteLine(resultP1);
        var clientConnection1 = clientProvider.CreateConnection();
        var productRows1 = await clientProvider.GetDatabaseBuilder().GetTableAsync("ProductCategory", null, clientConnection1);
        Console.WriteLine($"rows: {productRows1.Rows.Count}");


        var p2 = new SyncParameters { { "Color", "Silver" }, { "ProductCategoryID", "61b21b65-e16a-4be7-9300-4d8e9db861be" } };
        var resultP2 = await agent.SynchronizeAsync(setup, p2);
        Console.WriteLine(resultP2);

        var clientConnection2 = clientProvider.CreateConnection();
        var productRows2 = await clientProvider.GetDatabaseBuilder().GetTableAsync("ProductCategory", null, clientConnection2);
        Console.WriteLine($"rows: {productRows2.Rows.Count}");



**Test Results**

  **var p0 = new SyncParameters { { "Color", "Red" }, { "ProductCategoryID", "5515f857-075b-4f9a-87b7-43b4997077b3" } };**
Synchronization done.
        Total changes  uploaded: 0
        Total changes  downloaded: 59
        Total changes  applied on client: 59
        Total changes  applied on server: 0
        Total changes  failed to apply on client: 0
        Total changes  failed to apply on server: 0
        Total resolved conflicts: 0
        Total duration :00.00:01.991
rows: 42


**var p1 = new SyncParameters { { "Color", "Black" }, { "ProductCategoryID", "5515f857-075b-4f9a-87b7-43b4997077b3" } };**
Synchronization done.
        Total changes  uploaded: 0
        Total changes  downloaded: 53
        Total changes  applied on client: 53
        Total changes  applied on server: 0
        Total changes  failed to apply on client: 0
        Total changes  failed to apply on server: 0
        Total resolved conflicts: 0
        Total duration :00.00:01.772
rows: 42


  **var p2 = new SyncParameters { { "Color", "Silver" }, { "ProductCategoryID", "61b21b65-e16a-4be7-9300-4d8e9db861be" } };**
Synchronization done.
        Total changes  uploaded: 0
        Total changes  downloaded: 52
        Total changes  applied on client: 52
        Total changes  applied on server: 0
        Total changes  failed to apply on client: 0
        Total changes  failed to apply on server: 0
        Total resolved conflicts: 0
        Total duration :00.00:01.935
rows: 42



**REPEAT SAME FILTERS AGAIN**  ->  records for red and black will also be removed from the hub one after other. if a new ProductCategoryId is synced, last downloaded silver will also be removed.

  **var p0 = new SyncParameters { { "Color", "Red" }, { "ProductCategoryID", "5515f857-075b-4f9a-87b7-43b4997077b3" } };**
Synchronization done.
        Total changes  uploaded: 28
        Total changes  downloaded: 0
        Total changes  applied on client: 0
        Total changes  applied on server: 28
        Total changes  failed to apply on client: 0
        Total changes  failed to apply on server: 0
        Total resolved conflicts: 0
        Total duration :00.00:01.952
rows: 42

**var p1 = new SyncParameters { { "Color", "Black" }, { "ProductCategoryID", "5515f857-075b-4f9a-87b7-43b4997077b3" } };**
Synchronization done.
        Total changes  uploaded: 28
        Total changes  downloaded: 0
        Total changes  applied on client: 0
        Total changes  applied on server: 0
        Total changes  failed to apply on client: 0
        Total changes  failed to apply on server: 0
        Total resolved conflicts: 28
        Total duration :00.00:01.794
rows: 42

 **var p2 = new SyncParameters { { "Color", "Silver" }, { "ProductCategoryID", "61b21b65-e16a-4be7-9300-4d8e9db861be" } };**
Synchronization done.
        Total changes  uploaded: 11
        Total changes  downloaded: 0
        Total changes  applied on client: 0
        Total changes  applied on server: 0
        Total changes  failed to apply on client: 0
        Total changes  failed to apply on server: 0
        Total resolved conflicts: 11
        Total duration :00.00:01.792
rows: 42

**RECORDS REMAINED IN THE END** 
![image](https://github.com/Mimetis/Dotmim.Sync/assets/111710684/478ba004-5194-499e-83e7-59f5aefbe23f)



**CASCADE mentioned above** -  Cascade can be applied after the first sync, once the database is autogenerated.
![image](https://github.com/Mimetis/Dotmim.Sync/assets/111710684/56bf735e-deb4-49b5-bdc2-decc558daec1)

To Avoid Errors, 
- removed the ForeignKey reference to ProductID , from SalesOrderDetail  table 
- removed the ForeignKey ParentProductCategory referencing ProductCategory in same table.

please kindly let us know, if you want anything more.

Thanks
Sojan

**Comment by sojan1** [comment]
Hi @Mimetis 

Have you had a chance to look at the sample application made on  Adventureworks db script in the above post.
https://github.com/sojan1/Dotmim.Sync.SampleConsole/tree/master

**Reply by Mimetis** [reply]
Yes, but since your sample is not reproducing your error, I was waiting for a valid sample.

Seems you have a sqlite database that is not built using DMS on the first sync.
Seems your sqlite schema already exists before first sync.
Seems your sample does not provide any valid explanation or code to setup the first sync
Seems your sample does not do any delete after your sync, that will reproduce the error correctly.

So, NO, I did not had a chance to review your sample, as your sample is not reproducing the error.
Once I will have a valid sample reproducing your error that do not force me to invest hours to setup it, I will not try anything :)

**Reply by sojan1** [reply]
Hi @Mimetis ,

After **DMS creates SQLite advworks.db** , set a CASCADE in one table   -   **Only thing to do**
  SQLite -> advworks.db->Product (table)-> **ProductCategoryID** (field) ->  set cascade on delete 

**The Issue** : The ProductCategory table rows are re-downloading on every sync, when filter parameters are new

Everything about the **issue is shown in 4 minutes and 32 seconds video below**
https://sojan-geroc.tinytake.com/msc/ODc3MDc0MF8yMjE4NjI1OA


**Response to your comments**
--------------------------------------------------------------------------------------------------------------------------------
Seems you have a sqlite database that is not built using DMS on the first sync. - 
**No, DMS creates the SQLite advworks.db** 

Seems your sqlite schema already exists before first sync.- 
**No. DMS does everything, SQL schema doesn't exist first**.

Seems your sample does not provide any valid explanation or code to setup the first sync. - 
**There is no settings or code other than below.**

```

    var serverProvider = new SqlSyncProvider(DBHelper.GetDatabaseConnectionString(serverDbName));
    SqliteSyncProvider clientProvider = new SqliteSyncProvider("advworks.db");

    var setup = new SyncSetup("ProductCategory", "Product");
    var options = new SyncOptions();
    var agent = new SyncAgent(clientProvider, serverProvider, options);

    var proFilter = new SetupFilter("Product");
    proFilter.AddParameter("Color", "Product", false);
    proFilter.AddWhere("Color", "Product", "Color");
    proFilter.AddParameter("ProductCategoryID", "Product", false);
    proFilter.AddWhere("ProductCategoryID", "Product", "ProductCategoryID");
    setup.Filters.Add(proFilter);


    var p0 = new SyncParameters { { "Color", "Red" }, { "ProductCategoryID", "5515f857-075b-4f9a-87b7-43b4997077b3" } };
    var resultP0 = await agent.SynchronizeAsync(setup, p0);
```
   

Seems your sample does not do any delete after your sync, which will reproduce the error correctly. -  
**If i sync the red the second time, after syncing black or silver, red will be removed from the Hub**.
![image](https://github.com/Mimetis/Dotmim.Sync/assets/111710684/de08f211-2a36-4e12-8183-0694cbffea55)

Thanks

**Reply by Mimetis** [reply]
Here is the setup I used (and what I was waiting for, but seems you are not able to do it ...)

Based on **AdventureWorks.sql** script as server side using `SqlSyncProvider`:

- Using `SqliteSyncProvider` as client provider
- Creating manually the client database where `CASCADE DELETE` is set on the Foreign Key for `ProductCategoryId ` in `Product` table
- Using two filters to sync
- First filter sync **black color**
- Second filter sync **red color**, but all products for black color are deleted, only if `CASCADE DELETE` is set 

``` cs
var serverProvider = new SqlSyncProvider(DBHelper.GetDatabaseConnectionString(serverDbName));
var clientProvider = new SqliteSyncProvider(Path.GetRandomFileName().Replace(".", "").ToLowerInvariant() + ".db");

// ------------
// Create sqlite tables
// --------------

var commandText = @"
                    BEGIN TRANSACTION;
                    CREATE TABLE IF NOT EXISTS ""ProductCategory"" (
                        ""ProductCategoryID""	text NOT NULL COLLATE NOCASE,
                        ""ParentProductCategoryID""	text COLLATE NOCASE,
                        ""Name""	text NOT NULL COLLATE NOCASE,
                        ""rowguid""	text COLLATE NOCASE,
                        ""ModifiedDate""	text,
                        ""IsActive""	integer NOT NULL,
                        PRIMARY KEY(""ProductCategoryID"")
                    );
                    CREATE TABLE IF NOT EXISTS ""Product"" (
                        ""ProductID""	integer,
                        ""Name""	text NOT NULL COLLATE NOCASE,
                        ""ProductNumber""	text NOT NULL COLLATE NOCASE,
                        ""Color""	text COLLATE NOCASE,
                        ""StandardCost""	numeric NOT NULL,
                        ""ListPrice""	numeric NOT NULL,
                        ""Size""	text COLLATE NOCASE,
                        ""Weight""	numeric,
                        ""ProductCategoryID""	text COLLATE NOCASE,
                        ""ProductModelID""	integer,
                        ""SellStartDate""	text NOT NULL,
                        ""SellEndDate""	text,
                        ""DiscontinuedDate""	text,
                        ""ThumbNailPhoto""	blob,
                        ""ThumbnailPhotoFileName""	text COLLATE NOCASE,
                        ""rowguid""	text NOT NULL COLLATE NOCASE,
                        ""ModifiedDate""	text NOT NULL,
                        PRIMARY KEY(""ProductID""),
                        FOREIGN KEY(""ProductCategoryID"") REFERENCES ""ProductCategory""(""ProductCategoryID"")
                                                                                     ON DELETE CASCADE
                    );
                    COMMIT;";
// -- Create initial SQLITE database
await DBHelper.ExecuteSqliteScriptAsync(clientProvider.ConnectionString, commandText);

// Create sync
var setup = new SyncSetup("ProductCategory", "Product");
var options = new SyncOptions();

var agent = new SyncAgent(clientProvider, serverProvider, options);

var proFilter = new SetupFilter("Product");
proFilter.AddParameter("Color", "Product", false);
proFilter.AddWhere("Color", "Product", "Color");
proFilter.AddParameter("ProductCategoryID", "Product", false);
proFilter.AddWhere("ProductCategoryID", "Product", "ProductCategoryID");
setup.Filters.Add(proFilter);

// 1st sync
var pColorRed = new SyncParameters { { "Color", "Red" }, { "ProductCategoryID", "5515f857-075b-4f9a-87b7-43b4997077b3" } };
var resultColorRed = await agent.SynchronizeAsync(setup, pColorRed);
Console.WriteLine(resultColorRed);

// 2nd Sync
var pColorBlack = new SyncParameters { { "Color", "Black" }, { "ProductCategoryID", "5515f857-075b-4f9a-87b7-43b4997077b3" } };
var resultColorBlack = await agent.SynchronizeAsync(setup, pColorBlack);
Console.WriteLine(resultColorBlack);

```
As stated here, https://www.sqlite.org/lang_conflict.html:
 > When a [UNIQUE](https://www.sqlite.org/lang_createtable.html#uniqueconst) or [PRIMARY KEY](https://www.sqlite.org/lang_createtable.html#primkeyconst) constraint violation occurs, the REPLACE algorithm deletes pre-existing rows that are causing the constraint violation prior to inserting or updating the current row and the command continues executing normally.

When you are running the second sync, all **Product Categories** rows are downloaded again and a `INSERT OR REPLACE` statement is used.
Since `DELETE CASCADE` is on, all the previous **Products** are deleted, and then sync back to server (as deleted).

`RecursiveTriggers` (set to `false` / `true`) will not help as we don't care about the `DELETE` trigger, but we care about the `CASCADE DELETE`.

**BUT**, as stated in the documentation, it's a constraint failing case, that we can disable by using `PRAGMA foreign_keys = false;`

For this, I'm using the `SqliteConnectionStringBuilder` to set explicitly to turn off foreign keys constraints:

``` cs
var sqliteDatabaseName = Path.GetRandomFileName().Replace(".", "").ToLowerInvariant() + ".db";
var builder = new SqliteConnectionStringBuilder {
    DataSource = sqliteDatabaseName,
    ForeignKeys = false,
};
var clientProvider = new SqliteSyncProvider(builder);
```

And it's working as expected...

**BUT I'M NOT SURE ABOUT ANY SIDE EFFECTS WHEN DISABLING FOREIGN KEYS ON SQLITE.**

**My advice: Do Not Use Cascade Delete.**


PS : **As you can see**, I put a lot of efforts to make a clear message, with all the required steps to reproduce the error.
This post uses **Bold** statements and `Code` statements as well CSHARP code format for code paragraphs.
Everything is well formatted to understand correctly and have a clear view on the meaning of the post...

**I WOULD LOVE** to see such messages every time someone is raising a bug in DMS.

**Comment by sojan1** [comment]
Hi @Mimetis,

We really appreciate your time in investigating this and we are really glad that you were able to replicate the same behaviour we see.

Our question is, **why does this behaviour differ from 0.9.7** with the CASCADE on the database relationship still enabled?

**0.9.7 does not mark records for deletion** and does not subsequently delete records from the server.

PS: In the future we will use your existing database and samples to avoid confusion, and will do our best with formatting.

Thanks
Sojan

**Reply by Mimetis** [reply]
No idea

BTW, setting `DisableConstraintsOnApplyChanges` for SQLite will do the same

``` cs
var options = new SyncOptions
{
    DisableConstraintsOnApplyChanges = false
};

```
